src/ray/core_worker/core_worker.cc

```diff
@@ -2238,6 +2238,9 @@ void CoreWorker::HandleCancelTask(const rpc::CancelTaskRequest &request,
   reply->set_attempt_succeeded(success);
   send_reply_callback(Status::OK(), nullptr, nullptr);
 
+  // TODO: fix race condition to avoid using this hack
+  requested_task_running = main_thread_task_id_ == task_id;
+
```

Comment:
Maybe set this before the `send_reply_callback`?

Reply:
ok, let me try that!